source/common/config/watch_map_impl.cc

```diff
@@ -0,0 +1,171 @@
+#include "common/config/watch_map_impl.h"
+
+namespace Envoy {
+namespace Config {
+
+WatchPtr WatchMapImpl::addWatch(SubscriptionCallbacks& callbacks) {
+  auto watch = std::make_unique<Watch>(*this, callbacks);
+  wildcard_watches_.insert(watch.get());
+  watches_.insert(watch.get());
+  return watch;
+}
+
+void WatchMapImpl::removeWatch(Watch* watch) {
+  wildcard_watches_.erase(watch); // may or may not be in there, but we want it gone.
+  watches_.erase(watch);
+}
+
+AddedRemoved WatchMapImpl::updateWatchInterest(Watch* watch,
+                                               const std::set<std::string>& update_to_these_names) {
+  if (update_to_these_names.empty()) {
+    wildcard_watches_.insert(watch);
+  } else {
+    wildcard_watches_.erase(watch);
+  }
+
+  std::vector<std::string> newly_added_to_watch;
+  std::set_difference(update_to_these_names.begin(), update_to_these_names.end(),
+                      watch->resource_names_.begin(), watch->resource_names_.end(),
+                      std::inserter(newly_added_to_watch, newly_added_to_watch.begin()));
+
+  std::vector<std::string> newly_removed_from_watch;
+  std::set_difference(watch->resource_names_.begin(), watch->resource_names_.end(),
+                      update_to_these_names.begin(), update_to_these_names.end(),
+                      std::inserter(newly_removed_from_watch, newly_removed_from_watch.begin()));
+
+  watch->resource_names_ = update_to_these_names;
+
+  return AddedRemoved(findAdditions(newly_added_to_watch, watch),
```

Comment:
There's quite a bit of copying going on here I think. I don't want to micro-optimize, but could we at least make `AddedRemoved` take a `&&` move argument for the two sets?

Reply:
Yes, great idea, done.